File: samples/ios/token-refresh/Quickstart/AuthExpiry.swift
import Foundation

enum AuthExpiry {
    /// `.valid(remainingMs:)` — token is not expired. `remainingMs` is the number
    /// of milliseconds until expiry, or `nil` if no expiry was provided (treat as
    /// non-expiring; matches the RN sample's "no timer when no date" behavior).
    /// `.expired` — the provided expiry has already elapsed.
    enum Result: Equatable {
        case valid(remainingMs: Int?)
        case expired
    }

    static func evaluate(expiresAt: Date?) -> Result {
        guard let expiresAt else { return .valid(remainingMs: nil) }
        let remaining = expiresAt.timeIntervalSinceNow
        if remaining <= 0 { return .expired }
        return .valid(remainingMs: Int(remaining * 1000))
    }
}

File: samples/android/login-pkce/app/src/main/java/com/secureauth/quickstart/AuthExpiry.kt
package com.secureauth.quickstart

object AuthExpiry {
    /**
     * `Valid(remainingMs)` — token is not expired. `remainingMs` is the number of
     * milliseconds until expiry, or `null` if no expiry was provided (treat as
     * non-expiring; matches the iOS sample's "no timer when no date" behavior).
     * `Expired` — the provided expiry has already elapsed.
     */
    sealed class Result {
        data class Valid(val remainingMs: Long?) : Result()
        data object Expired : Result()
    }

    fun evaluate(expiresAt: Long?): Result {
        if (expiresAt == null) return Result.Valid(null)
        val remaining = expiresAt - System.currentTimeMillis()
        return if (remaining <= 0L) Result.Expired else Result.Valid(remaining)
    }
}
